plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/pvc/CommonPVCStrategyWithEphemeralWorkspacesSupport.java

```diff
@@ -0,0 +1,164 @@
+        }
+      }
+    } else {
+      super.provision(k8sEnv, identity);
+    }
+  }
+
+  @Override
+  public void cleanup(String workspaceId) throws InfrastructureException {
+    if (!isEphemeral(workspaceId)) {
+      super.cleanup(workspaceId);
+    }
+  }
+
+  @Override
+  public void prepare(KubernetesEnvironment k8sEnv, String workspaceId)
+      throws InfrastructureException {
+    if (!isEphemeral(workspaceId)) {
+      super.prepare(k8sEnv, workspaceId);
+    }
+  }
+
+  private void addMachineVolumes(
+      String workspaceId, Pod pod, Container container, Map<String, Volume> volumes) {
+    if (volumes.isEmpty()) {
+      return;
+    }
+    for (Entry<String, Volume> volumeEntry : volumes.entrySet()) {
+      final String volumePath = volumeEntry.getValue().getPath();
+      final String volumeName =
+          LOGS_VOLUME_NAME.equals(volumeEntry.getKey())
+              ? volumeEntry.getKey() + '-' + pod.getMetadata().getName()
+              : volumeEntry.getKey();
+      final String uniqueVolumeMountName = Names.generateName(volumeNamePrefix + '-');
+      container
+          .getVolumeMounts()
+          .add(
+              newVolumeMount(
+                  uniqueVolumeMountName,
+                  volumePath,
+                  getSubPath(workspaceId, volumeName, Names.machineName(pod, container))));
+      addEmptyDirVolumeIfAbsent(pod.getSpec(), uniqueVolumeMountName);
+    }
+  }
+
+  private void addEmptyDirVolumeIfAbsent(PodSpec podSpec, String uniqueVolumeMountName) {
+    if (podSpec
+        .getVolumes()
+        .stream()
+        .noneMatch(volume -> volume.getName().equals(uniqueVolumeMountName))) {
+      podSpec
+          .getVolumes()
+          .add(
+              new VolumeBuilder()
+                  .withName(uniqueVolumeMountName)
+                  .withNewEmptyDir()
+                  .endEmptyDir()
+                  .build());
+    }
+  }
+
+  private String getSubPath(String workspaceId, String volumeName, String machineName) {
+    // logs must be located inside the folder related to the machine because
+    // few machines can
+    // contain the identical agents and in this case, a conflict is
+    // possible.
+    if (LOGS_VOLUME_NAME.equals(volumeName)) {
+      return workspaceId + '/' + volumeName + '/' + machineName;
+    }
+    return workspaceId + '/' + volumeName;
+  }
+
+  /**
+   * @param workspaceId
+   * @return true if workspace config contains `mountSources` attribute which is set to true, false
+   *     otherwise
+   */
+  private boolean isEphemeral(String workspaceId) {
+    try {
+      WorkspaceImpl workspace = workspaceManager.getWorkspace(workspaceId);
+      String mountSources = workspace.getConfig().getAttributes().get(MOUNT_SOURCES_ATTRIBUTE);
+      return !Boolean.parseBoolean(mountSources);
+    } catch (NotFoundException | ServerException e) {
+      throw new RuntimeException("Failed to load workspace info" + e.getMessage(), e);
```

Comment:
I think that any runtime exceptions in this area are discouraged. Please, use InfrastructureException or InternalInfrastructureException (specifically in this case) instead.